### ISSUE TITLE
How to use multi-thread in C api?

### ISSUE THREAD

**Issue by hmchen-github** [dmlc/xgboost#8414]
I am using `XGDMatrixCreateFromMat` to create the DMatrix, `XGBoosterCreate` to create the model and `XGBoosterPredict` to predict the data. I don't know where to set the nthread to make the whole processing faster. 

For the `XGDMatrixCreateFromMat`, 
1. I saw it is hard-coded [nthread=1 in the implementation](https://github.com/dmlc/xgboost/blob/master/src/c_api/c_api.cc#L457). Is that intentional? 
2. Does multithread make a difference there?

For the `XGBoosterCreate` and `XGBoosterPredict`, I don't see any place to set the nthread. 
1. Is the default max number of threads?
3. Should I use `XGBoosterSetParam` to it? Any example or documentation?

**Comment by hcho3** [comment]
> I saw it is hard-coded [nthread=1 in the implementation](https://github.com/dmlc/xgboost/blob/master/src/c_api/c_api.cc#L457). Is that intentional?

You can use `XGDMatrixCreateFromMat_omp` to enable multi-threading.

> For the XGBoosterCreate and XGBoosterPredict, I don't see any place to set the nthread.

You can call `XGBoosterSetParam` to set the `nthread` hyperparameter.

**Comment by hmchen-github** [comment]
Thanks for the quick response! For the `XGDMatrixCreateFromMat_omp`, should I use a specific number or just use 0 for max number of threads?

Same thing for the XGBoosterSetParam. Does nthread=0 mean max threads?

**Comment by hcho3** [comment]
Yes, nthread=0 will use max number of threads. However, in some systems, setting a lower nthread may produce better performance. Your mileage may vary.

**Comment by trivialfis** [comment]
I think you can see how to specify parameters for dmatrix construction with this demo: https://github.com/dmlc/xgboost/blob/b1b2524dbb6dd2216aaf1a46366728484d6c7717/demo/c-api/inference/inference.c#L120 

As for booster parameters, `XGBoosterSetParam` should be the function for the job.

**Comment by hmchen-github** [comment]
I modified the demo/c-api example and added XGBoosterSetParam(booster, "nthread", "4") in

https://github.com/dmlc/xgboost/blob/b1b2524dbb6dd2216aaf1a46366728484d6c7717/demo/c-api/inference/inference.c#L143

And use clock() to measure the processing time of the `XGBoosterPredictFromDMatrix` and `XGBoosterPredictFromDense`. I also changed the `#define N_SAMPLES 128` to a much larger number to make it run long enough.

But I didn't see any difference on the running time, for either nthread=4, 8, 16 or without SetParam.

I feel the nthread is not actually set. Did I miss anything?

**Comment by trivialfis** [comment]
Try to make the sample size larger, let's start with 10 thousands, and observe htop. we assign blocks of sample (instead of one sample per thr ead) to each thread to have better batch performance.

**Comment by hmchen-github** [comment]
Yes, I used `#define N_SAMPLES 25600000`, which I think is large enough? The overall prediction time on my machine is around 6s. But that duration is very consistent for nthread=1,4,8,16. My expectation is that it will be shorter for more threads or longer for less threads.

Yes, I also used htop but I don't see any difference on the number of threads when changing nthread, that's why I think the nthread is not actually used.

**Comment by trivialfis** [comment]
> Yes, I also used htop but I don't see any difference on the number of threads when changing nthread

Hmm, so how many threads are you observing?

Also, could you please share the code that you are using for test?

**Comment by hmchen-github** [comment]
Never mind. I used clock() to measure the cpu usage so multithread will also increase the usage. I tried time() to measure the actual wall time and it did show multi-thread speed up the process. Using htop does not show extra threads under the process of the executive. But I did see more cores are busy when use multithreads. So I think it verifies that multithreads works.

Here is my test code if you or others are interested in. https://github.com/dmlc/xgboost/compare/master...hmchen-github:xgboost:test-nthread

Another question is about how many threads I should use. We used v0.6 before and recently upgraded to v1.5 for inference. The reason I am looking at this issue is that the data on v0.6 runs slower on v1.5. So I guess the threads might be different. Do you know how many threads are used as default in v0.6?

**Comment by trivialfis** [comment]
It depends on your data. XGBoost is not good at running inference with small number of data. For that you need specialized inference libraries.

Also, there are inplcae prediction in latest XGBoost, which avoids construction of DMatrix altogether and might help increase both throughput and latency.

Lastly, prediction result is stored in thread local memory. You can control the thread outside and run small batches of prediction using the thread allocated by you instead of openmp in xgboost.
